src/engine/strat_engine/backstore/backstore.rs

```diff
@@ -215,96 +456,373 @@ impl Backstore {
-        self.data_tier
-            .dm_device
-            .teardown(dm)
+        match self.cache {
+                Some(cache) => cache.teardown(dm),
+                None => self.linear.expect("self.cache.is_none()").teardown(dm),
+            }
             .map_err(|e| e.into())
     }
 
     /// Return the device that this tier is currently using.
-    /// WARNING: This may change it the backstore switches between its
-    /// cache and its non-cache incarnations, among other reasons.
+    /// This changes, depending on whether the backstore is supporting a cache
+    /// or not.
     pub fn device(&self) -> Device {
-        self.data_tier.dm_device.device()
+        self.cache
+            .as_ref()
+            .map(|d| d.device())
+            .or_else(|| self.linear.as_ref().map(|d| d.device()))
+            .expect("must be one or the other")
     }
 
-
     /// Lookup an immutable blockdev by its Stratis UUID.
-    pub fn get_blockdev_by_uuid(&self, uuid: DevUuid) -> Option<&BlockDev> {
-        self.data_tier.block_mgr.get_blockdev_by_uuid(uuid)
+    pub fn get_blockdev_by_uuid(&self, uuid: DevUuid) -> Option<(BlockDevTier, &BlockDev)> {
+        self.data_tier
+            .get_blockdev_by_uuid(uuid)
+            .or_else(|| {
+                         self.cache_tier
+                             .as_ref()
+                             .and_then(|c| c.get_blockdev_by_uuid(uuid))
+                     })
     }
 
     /// Lookup a mutable blockdev by its Stratis UUID.
-    pub fn get_mut_blockdev_by_uuid(&mut self, uuid: DevUuid) -> Option<&mut BlockDev> {
-        self.data_tier.block_mgr.get_mut_blockdev_by_uuid(uuid)
+    pub fn get_mut_blockdev_by_uuid(&mut self,
+                                    uuid: DevUuid)
+                                    -> Option<(BlockDevTier, &mut BlockDev)> {
+        let cache_tier = &mut self.cache_tier;
+        self.data_tier
+            .get_mut_blockdev_by_uuid(uuid)
+            .or_else(move || {
+                         cache_tier
+                             .as_mut()
+                             .and_then(|c| c.get_mut_blockdev_by_uuid(uuid))
+                     })
     }
 
     /// The number of sectors in the backstore given up to Stratis
-    /// metadata.
-    pub fn metadata_size(&self) -> Sectors {
+    /// metadata on devices in the data tier.
+    pub fn datadev_metadata_size(&self) -> Sectors {
         self.data_tier.metadata_size()
     }
 
-    /// Write the given data directly to the blockdevs that make up the
-    /// backstore. This action bypasses the DM device entirely,
-    /// in order to allow control over which blockdevs the metadata is written
-    /// to.
-    pub fn save_state(&mut self, metadata: &[u8]) -> EngineResult<()> {
-        self.data_tier.block_mgr.save_state(metadata)
+    /// Write the given data to the data tier's devices.
+    pub fn datadev_save_state(&mut self, metadata: &[u8]) -> EngineResult<()> {
```

Comment:
suggest rename back to save_state. The fact that we're just saving state to our data tier blockdevs is an implementation detail.

Reply:
Ok!